libs/sdk-utils/core/src/lib/wrapper.ts

```diff
@@ -0,0 +1,37 @@
+import { Contract, ContractInterface } from 'ethers';
+import { SignerOrProvider, typesafeContract } from './utils';
+
+/**
+ * ContractWrapper is an abstract class for implementing the decorator pattern on top of the ethers contract instance.
+ * The additional functionality should be passed as the generic `A`, and when the `wrap` method is called, a decorated
+ * version of the original contract will be returned.
+ *
+ * This can be used as a normal ethers contract anywhere in the applicaiton.
+ *
+ * @param A The additional wrapped functionality, it should be an interface or object type but can contain any number of properties or methods.
+ **/
+export abstract class ContractWrapper<A> {
```

Comment:
I really like this abstraction although the name is a bit misleading as this won't wrap anything, it is an implementation of the factory method pattern, I think it should be named appropriately

Reply:
ContractDecoratorFactory?

Reply:
Oh now...this is quickly becoming Java-like 😂